lib/src/main/java/oiccli/Service.java

```diff
@@ -0,0 +1,443 @@
+        if (method == null) {
+            method = this.httpMethod;
+        }
+
+        if (requestArgs == null) {
+            requestArgs = new HashMap<>();
+        }
+
+        Map<String, String> hMap = new HashMap<>();
+        for (String key : args.keySet()) {
+            if (args.get(key) != null && !specialArgs.contains(key)) {
+                hMap.put(key, args.get(key));
+            }
+        }
+
+        Message cis = this.constructMessage(clientInfo, requestArgs, args);
+
+        if (this.events != null) {
+            this.events.store("Protocol request", cis);
+        }
+
+        cis.setLax(lax);
+        Map<String, String> hArg = new HashMap<>();
+        if (StringUtil.isNotNullAndNotEmpty(authenticationMethod)) {
+            hArg = this.initAuthenticationMethod(cis, clientInfo, authenticationMethod, requestArgs, args);
+        }
+
+        if (hArg != null) {
+            if (args.containsKey("headers")) {
+                args.get("headers").update(hArg.get("headers"));
+            } else {
+                args.put("headers", hArg.get("headers"));
+            }
+        }
+
+        if (bodyType.equals("json")) {
+            args.put("contentType", "application/json");
+        }
+
+        return this.uriAndBody(cis, method, args);
+    }
+
+    public Map<String, Object> updateHttpArgs(Map<String, String> httpArgs, Map<String, Object> info) {
+        Map<String, String> hArgs = info.get("hArgs");
+        if (hArgs == null) {
+            hArgs = new HashMap<>();
+        }
+
+        if (httpArgs == null) {
+            httpArgs = hArgs;
+        } else {
+            httpArgs.update(info.get("h_args"));
+        }
+
+        info.put("httpArgs", httpArgs);
+        return info;
+    }
+
+    public Map<String, Object> doRequestInit(ClientInfo clientInfo, String bodyType, String method, String authenticationMethod,
+                                             Map<String, Object> requestArgs, Map<String, String> httpArgs, Map<String, String> args) throws NoSuchFieldException, IllegalAccessException, MissingEndpoint, UnsupportedEncodingException, UnsupportedType {
+        if (!StringUtil.isNotNullAndNotEmpty(method)) {
+            method = this.httpMethod;
+        }
+        if (!StringUtil.isNotNullAndNotEmpty(authenticationMethod)) {
+            authenticationMethod = this.defaultAuthenticationMethod;
+        }
+        if (!StringUtil.isNotNullAndNotEmpty(bodyType)) {
+            bodyType = this.bodyType;
+        }
+
+        Map<String, Object> info = this.requestInfo(clientInfo, method, requestArgs, bodyType, authenticationMethod, false, args);
+
+        return this.updateHttpArgs(httpArgs, info);
+    }
+
+    public String getUrlInfo(String info) throws URISyntaxException {
+        String query = null, fragment = null;
+        if (info.contains("?") || info.contains("#")) {
+            URI uri = new URI(info);
+            query = uri.getQuery();
+            fragment = uri.getFragment();
```

Comment:
this returns null when fragment is undefined. https://docs.oracle.com/javase/7/docs/api/java/net/URI.html#getFragment()